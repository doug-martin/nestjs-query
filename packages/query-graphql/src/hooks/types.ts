export enum HookTypes {
  BEFORE_CREATE_ONE = 'BeforeCreateOne',
  BEFORE_CREATE_MANY = 'BeforeCreateMany',
  BEFORE_UPDATE_ONE = 'BeforeUpdateOne',
  BEFORE_UPDATE_MANY = 'BeforeUpdateMany',
  BEFORE_DELETE_ONE = 'BeforeDeleteOne',
  BEFORE_DELETE_MANY = 'BeforeDeleteMany',
  BEFORE_QUERY_MANY = 'BeforeQueryMany',
  BEFORE_FIND_ONE = 'BeforeFindOne',
}
