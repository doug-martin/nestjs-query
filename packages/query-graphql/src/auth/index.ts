export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
