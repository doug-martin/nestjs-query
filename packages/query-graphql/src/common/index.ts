export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
