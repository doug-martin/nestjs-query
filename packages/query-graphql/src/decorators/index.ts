export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
