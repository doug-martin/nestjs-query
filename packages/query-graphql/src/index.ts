export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
