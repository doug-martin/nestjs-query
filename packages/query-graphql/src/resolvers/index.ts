export { CRUDResolverOpts, CRUDResolver } from './crud.resolver';
export { CreateResolver, CreateResolverOpts } from './create.resolver';
export { ReadResolver, ReadResolverOpts } from './read.resolver';
export { UpdateResolver, UpdateResolverOpts } from './update.resolver';
export { DeleteResolver, DeleteResolverOpts } from './delete.resolver';
export { ReferenceResolver, ReferenceResolverOpts } from './reference.resolver';
export { Relatable, ResolverRelation, ResolverRelationReference } from './relations';
export { ResolverOpts } from './resolver.interface';
export { FederationResolver } from './federation';
