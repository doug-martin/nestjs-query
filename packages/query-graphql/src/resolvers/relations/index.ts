export { Relatable } from './relations.resolver';
export { ReadRelationsResolver } from './read-relations.resolver';
export { UpdateRelationsResolver } from './update-relations.resolver';
export { RemoveRelationsResolver } from './remove-relations.resolver';
export { ReferencesRelationsResolver } from './references-relation.resolver';
export { AggregateRelationsResolver } from './aggregate-relations.resolver';
export {
  RelationTypeMap,
  ReferencesOpts,
  ResolverRelationReference,
  RelationsOpts,
  ResolverRelation,
} from './relations.interface';
