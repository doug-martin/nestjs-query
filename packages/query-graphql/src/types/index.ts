export {
  PageInfoType,
  EdgeType,
  OffsetPageInfoType,
  ConnectionType,
  ConnectionOptions,
  CursorConnectionType,
  CursorConnectionOptions,
  ArrayConnectionType,
  ArrayConnectionOptions,
  OffsetConnectionType,
  OffsetConnectionOptions,
  StaticConnectionType,
  InferConnectionTypeFromStrategy,
} from './connection';
export { MutationArgsType } from './mutation-args.type';
export {
  FilterType,
  UpdateFilterType,
  DeleteFilterType,
  SubscriptionFilterType,
  QueryArgsType,
  QueryArgsTypeOpts,
  PagingTypes,
  CursorQueryArgsType,
  OffsetQueryArgsType,
  NonePagingQueryArgsType,
  CursorPagingType,
  OffsetPagingType,
  PagingStrategies,
} from './query';
export { CreateManyInputType } from './create-many-input.type';
export { CreateOneInputType } from './create-one-input.type';
export { UpdateManyResponseType } from './update-many-response.type';
export { UpdateOneInputType } from './update-one-input.type';
export { UpdateManyInputType } from './update-many-input.type';
export { DeleteManyResponseType } from './delete-many-reponse.type';
export { DeleteOneInputType } from './delete-one-input.type';
export { DeleteManyInputType } from './delete-many-input.type';
export { ConnectionCursorType, ConnectionCursorScalar } from './cursor.scalar';
export { RelationsInputType } from './relations-input.type';
export { RelationInputType } from './relation-input.type';
export { SubscriptionArgsType } from './subscription-args.type';
export { SubscriptionFilterInputType } from './subscription-filter-input.type';
export { AggregateResponseType, AggregateResponseOpts, AggregateArgsType } from './aggregate';
export { FindOneArgsType } from './find-one-args.type';
