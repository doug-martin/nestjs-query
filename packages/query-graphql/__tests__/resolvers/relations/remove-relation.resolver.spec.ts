import { when, deepEqual } from 'ts-mockito';
import { Resolver, Query } from '@nestjs/graphql';
import { RelationsOpts, RemoveRelationsResolver } from '../../../src/resolvers/relations';
import { RelationInputType, RelationsInputType } from '../../../src/types';
import {
  generateSchema,
  createResolverFromNest,
  TestResolverDTO,
  TestService,
  TestRelationDTO,
} from '../../__fixtures__';

@Resolver(() => TestResolverDTO)
class TestResolver extends RemoveRelationsResolver(TestResolverDTO, {
  one: { relation: { DTO: TestRelationDTO }, custom: { DTO: TestRelationDTO, relationName: 'other' } },
  many: { relations: { DTO: TestRelationDTO }, customs: { DTO: TestRelationDTO, relationName: 'others' } },
}) {
  constructor(service: TestService) {
    super(service);
  }
}

describe('RemoveRelationsResolver', () => {
  const expectResolverSDL = async (opts?: RelationsOpts) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends RemoveRelationsResolver(TestResolverDTO, opts ?? {}) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }

    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };
  it('should not add remove methods if one and many are empty', () => expectResolverSDL());

  describe('one', () => {
    it('should use the object type name', () => expectResolverSDL({ one: { relation: { DTO: TestRelationDTO } } }));

    it('should use the dtoName if provided', () =>
      expectResolverSDL({ one: { relation: { DTO: TestRelationDTO, dtoName: 'Test' } } }));

    it('should not add remove methods if disableRemove is true', () =>
      expectResolverSDL({ one: { relation: { DTO: TestRelationDTO, disableRemove: true } } }));

    it('should call the service findRelation with the provided dto and correct relation name', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const input: RelationInputType = {
        id: 'record-id',
        relationId: 'relation-id',
      };
      const output: TestResolverDTO = {
        id: 'record-id',
        stringField: 'foo',
      };
      when(mockService.removeRelation('relation', input.id, input.relationId, undefined)).thenResolve(output);
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.removeRelationFromTestResolverDTO({ input });
      return expect(result).toEqual(output);
    });

    it('should call the service findRelation with the provided dto and custom relation name', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const input: RelationInputType = {
        id: 'record-id',
        relationId: 'relation-id',
      };
      const output: TestResolverDTO = {
        id: 'record-id',
        stringField: 'foo',
      };
      when(mockService.removeRelation('other', input.id, input.relationId, undefined)).thenResolve(output);
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.removeCustomFromTestResolverDTO({ input });
      return expect(result).toEqual(output);
    });
  });

  describe('many', () => {
    it('should use the object type name', () => expectResolverSDL({ many: { relations: { DTO: TestRelationDTO } } }));

    it('should use the dtoName if provided', () =>
      expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, dtoName: 'Test' } } }));

    it('should not add remove many methods if disableRemove is true', () =>
      expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, disableRemove: true } } }));

    it('should call the service findRelation with the provided dto and correct relation name', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const input: RelationsInputType = {
        id: 'id-1',
        relationIds: ['relation-id-1', 'relation-id-2'],
      };
      const output: TestResolverDTO = {
        id: 'record-id',
        stringField: 'foo',
      };
      when(mockService.removeRelations('relations', input.id, deepEqual(input.relationIds), undefined)).thenResolve(
        output,
      );
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.removeRelationsFromTestResolverDTO({ input });
      return expect(result).toEqual(output);
    });

    it('should call the service findRelation with the provided dto and correct custom relation name', async () => {
      const { resolver, mockService } = await createResolverFromNest(TestResolver);
      const input: RelationsInputType = {
        id: 'id-1',
        relationIds: ['relation-id-1', 'relation-id-2'],
      };
      const output: TestResolverDTO = {
        id: 'record-id',
        stringField: 'foo',
      };
      when(mockService.removeRelations('others', input.id, deepEqual(input.relationIds), undefined)).thenResolve(
        output,
      );
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-call
      const result = await resolver.removeCustomsFromTestResolverDTO({ input });
      return expect(result).toEqual(output);
    });
  });
});
