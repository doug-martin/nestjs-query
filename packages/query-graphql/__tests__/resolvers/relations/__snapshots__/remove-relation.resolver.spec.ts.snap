// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveRelationsResolver many should not add remove many methods if disableRemove is true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`RemoveRelationsResolver many should use the dtoName if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  removeTestsFromTestResolverDTO(input: RemoveTestsFromTestResolverDTOInput!): TestResolverDTO!
}

input RemoveTestsFromTestResolverDTOInput {
  """The id of the record."""
  id: ID!

  """The ids of the relations."""
  relationIds: [ID!]!
}

`;

exports[`RemoveRelationsResolver many should use the object type name 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  removeRelationsFromTestResolverDTO(input: RemoveRelationsFromTestResolverDTOInput!): TestResolverDTO!
}

input RemoveRelationsFromTestResolverDTOInput {
  """The id of the record."""
  id: ID!

  """The ids of the relations."""
  relationIds: [ID!]!
}

`;

exports[`RemoveRelationsResolver one should not add remove methods if disableRemove is true 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

`;

exports[`RemoveRelationsResolver one should use the dtoName if provided 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  removeTestFromTestResolverDTO(input: RemoveTestFromTestResolverDTOInput!): TestResolverDTO!
}

input RemoveTestFromTestResolverDTOInput {
  """The id of the record."""
  id: ID!

  """The id of relation."""
  relationId: ID!
}

`;

exports[`RemoveRelationsResolver one should use the object type name 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

type Mutation {
  removeRelationFromTestResolverDTO(input: RemoveRelationFromTestResolverDTOInput!): TestResolverDTO!
}

input RemoveRelationFromTestResolverDTOInput {
  """The id of the record."""
  id: ID!

  """The id of relation."""
  relationId: ID!
}

`;

exports[`RemoveRelationsResolver should not add remove methods if one and many are empty 1`] = `
type TestResolverDTO {
  id: ID!
  stringField: String!
}

type Query {
  test: TestResolverDTO!
}

`;
